### PR TITLE
Dockerized userspace emulator that doesn't rely on binfmts

### PR DESCRIPTION
**Pull request by tsarn**
Slightly modified this patch for QEMU and got it to work with Docker: https://patchwork.kernel.org/patch/8052841/

### CODE REVIEW

**Comment by AntonKozlov** [comment]
(updated: hit send too early)

I like the approach, and how fearlessly you took it. Cool! 

0. What about a qemu project fork on github instead of maintaining a patch? I think you can create repos in `azul-reaseach` project. 

1. Have you considered to upstream these to qemu? 

2. What about building all qemu arches once and providing them as a base image? 

3. Is there any way this approach can be used for something like https://ownyourbits.com/2018/06/27/running-and-building-arm-docker-containers-in-x86/?

3 -- may be not on our critical path for RISC-V, but it may bring this contribution in a broader context and make it profitable for more people

**Comment by tsarn** [comment]
> 1\. Have you considered to upstream these to qemu?

Well I just took someone else's patch that didn't make into qemu and fixed it so recursive `execve()` calls would work too. If it wasn't accepted then, I doubt it would be accepted now.

> 2\. What about building all qemu arches once and providing them as a base image?

Do you mean system to build docker images like this one for different architectures? Sounds pretty interesting.

> 3\. Is there any way this approach can be used for something like https://ownyourbits.com/2018/06/27/running-and-building-arm-docker-containers-in-x86/?

It's possible to copy `qemu` and use `SHELL` command in Dockerfile to forward `RUN` commands through it. Unfortunately I can't find a way to override exec form of `RUN` (something like `RUN ["ls", "-a"]`). For `CMD` there exists the `ENTRYPOINT` command that does what I want, but I apparently not for `RUN`.

**Comment by AntonKozlov** [comment]
> Well I just took someone else's patch that didn't make into qemu and fixed it so recursive `execve()` calls would work too. If it wasn't accepted then, I doubt it would be accepted now.

Is there any discussion? It's likely there is specific issues that probably haven't addressed. What if your recursive patch fixes those?

I'm just trying to understand how good is the patch. Anyway, I think we'll use it for now. The question is how. If we think the patch will not change at all, I'm OK to store it as patch. But if we expect more changes or you try to upstream it, I suggest use a qemu fork and apply the patch there.

> > 2. What about building all qemu arches once and providing them as a base image?
> 
> Do you mean system to build docker images like this one for different architectures? Sounds pretty interesting.

I'm not sure about `system to build images`. I mean a docker image (for x86) that will have all arches (riscv, arm64/32, ...) built with the patch. If we push that to docker hub, then we'll avoid rebuilding qemu stuff for image with cross-compilers, tools,... The image with crosscompilers would then use image with qemu's as a base image in docker sense.

> > 3. Is there any way this approach can be used for something like https://ownyourbits.com/2018/06/27/running-and-building-arm-docker-containers-in-x86/?
> 
> It's possible to copy `qemu` and use `SHELL` command in Dockerfile to forward `RUN` commands through it. Unfortunately I can't find a way to override exec form of `RUN` (something like `RUN ["ls", "-a"]`). For `CMD` there exists the `ENTRYPOINT` command that does what I want, but I apparently not for `RUN`.

I'm looking for a way to build a non-x86 docker image from a container on x86 host (read "by Docker Hub automation")

Not supporting exec form of RUN looks fine for me. 

Docker have support for multiarch images. https://hub.docker.com/_/ubuntu -- when I do `docker run ubuntu:latest` on x86 and arm32 machines I get different images, each specialized to the CPU. 

If I want to extend arm32 image, I have to use `docker build` on arm32 system or set `binfmt`. But I'm interested in something like 
(pseudo-Dockerfile):
```
FROM qemu-users AS qemu

FROM ubuntu:latest-arm32 
COPY --from qemu-user /opt/qemu-arm /opt/qemu-arm

RUN /opt/qemu-arm --execve /opt/qemu apt-get install some-arm32-package
RUN rm /opt/qemu-arm # will not affect image size, of course
```
And being able to `docker build` this Dockerfile on x86 host and get arm32 image.

Would that be possible?

BTW --execve seems also to take path to qemu itself. Shouldn't it be a flag like `--execve-self`? (an altered semantic for execve on systems without binfmt support)

**Comment by tsarn** [comment]
I forked qemu at https://github.com/azul-research/qemu and reimplemented the patch myself and fixed a few issues that were discussed.

> BTW --execve seems also to take path to qemu itself. Shouldn't it be a flag like --execve-self? (an altered semantic for execve on systems without binfmt support)

I think taking path to itself could allow for more flexibility in some edge cases (or maybe not :) and doesn't require to lookup full path to qemu binary (which could be symlinked or something else that I've not thought of).


> I'm not sure about `system to build images`. I mean a docker image (for x86) that will have all arches (riscv, arm64/32, ...) built with the patch. If we push that to docker hub, then we'll avoid rebuilding qemu stuff for image with cross-compilers, tools,... The image with crosscompilers would then use image with qemu's as a base image in docker sense.

That is definitely doable.

> Docker have support for multiarch images

I'll look into it as I've never used it before.

> Would that be possible?

I don't see why not.

File: dev-riscv/emulator-userspace/scripts/configure.sh [hunk 1]
```diff
@@ -4,5 +4,9 @@ set -e
...
 rootfs="/riscv"
 
-chroot $rootfs apt-get update
-chroot $rootfs apt-get install -y build-essential git vim gdb
+run() {
+    chroot "$rootfs" /usr/bin/qemu-riscv64-static -e /usr/bin/qemu-riscv64-static "$@"
+}
+
+run /usr/bin/apt-get update
+run /usr/bin/apt-get install -y build-essential git vim gdb
```
Comment (AntonKozlov) [inline dev-riscv/emulator-userspace/scripts/configure.sh]: I got next error
```
The following packages have unmet dependencies:
 vim : Depends: libcanberra0 (>= 0.2) but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
The command '/bin/sh -c /configure.sh' returned a non-zero code: 100
```
I propose to remove `vim` for a while

Comment (tsarn) [reply dev-riscv/emulator-userspace/scripts/configure.sh]: Yeah, sadly this is what you get with Debian unstable :(
I'll remove it now.

**Comment by AntonKozlov** [comment]
> I forked qemu at https://github.com/azul-research/qemu and reimplemented the patch myself and fixed a few issues that were discussed.

Thanks! I'll setup docker hub automation shortly.

> >  --execve-self
> 
> I think taking path to itself could allow for more flexibility in some edge cases (or maybe not :)

That's true. Just thought as it may too many flexibility, obscuring most common use case.

> and doesn't require to lookup full path to qemu binary (which could be symlinked or something else that I've not thought of).

There is a trick: `/proc/self/exe` on Linux is always a symlink to actual library.

> > Docker have support for multiarch images
>
> I'll look into it as I've never used it before.

But please don't spend too much time for this, this would be nice, but it's certainly not required for the project.

**Pull request merged**